compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java

```diff
@@ -164,62 +116,123 @@ private STToken readTokenInReTerm() {
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.ASTERISK_TOKEN);
+            case LexerTerminals.PLUS:
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.PLUS_TOKEN);
+            case LexerTerminals.QUESTION_MARK:
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.QUESTION_MARK_TOKEN);
             case LexerTerminals.BACKSLASH:
-                boolean processed = processOnlyNumericEscapeOrControlEscape();
-                if (processed) {
-                    startMode(ParserMode.RE_QUANTIFIER);
-                    return getRegExpText(SyntaxKind.RE_ESCAPE);
+                // This is used to identify `\-` in ReCharSetAtomNoDash. It's an invalid node if it's in any other
+                // position.
+                if (this.reader.peek(1) == LexerTerminals.MINUS) {
+                    this.reader.advance(2);
+                    return getRegExpSyntaxToken(SyntaxKind.ESCAPED_MINUS_TOKEN);
                 }
-                this.reader.advance();
-                return getRegExpSyntaxToken(SyntaxKind.BACK_SLASH_TOKEN);
-            // Handle "[" ["^"] [ReCharSet] "]".
+                return processReEscape();
+            // Start parsing character class [[^] [ReCharSet]].
             case LexerTerminals.OPEN_BRACKET:
                 this.reader.advance();
-                if (peek() == LexerTerminals.BITWISE_XOR) {
-                    startMode(ParserMode.RE_NEGATED_CHAR_CLASS_START);
-                } else {
-                    startMode(ParserMode.RE_CHAR_SET);
-                }
                 return getRegExpSyntaxToken(SyntaxKind.OPEN_BRACKET_TOKEN);
-            // Handle "(" ["?" ReFlagsOnOff ":"] ReDisjunction ")".
+            case LexerTerminals.CLOSE_BRACKET:
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.CLOSE_BRACKET_TOKEN);
+            case LexerTerminals.OPEN_BRACE:
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.OPEN_BRACE_TOKEN);
+            case LexerTerminals.CLOSE_BRACE:
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.CLOSE_BRACE_TOKEN);
+            // Start parsing capturing group ([? ReFlagsOnOff :] ReDisjunction).
             case LexerTerminals.OPEN_PARANTHESIS:
-                if (this.reader.peek(1) == LexerTerminals.QUESTION_MARK) {
-                    startMode(ParserMode.RE_FLAGS_START);
+                this.reader.advance();
+                if (peek() == LexerTerminals.QUESTION_MARK) {
+                    // Start a new parser mode to handle the flags.
+                    startMode(ParserMode.RE_FLAG_EXPRESSION);
                 } else {
-                    startMode(ParserMode.RE_CAPTURING_GROUP_RE_DISJUNCTION);
+                    startMode(ParserMode.RE_DISJUNCTION);
```

Comment:
Do we need to start the same mode again since we are already in the `RE_DISJUNCTION`?